ocaml/xapi/xapi_xenops.ml

```diff
@@ -647,29 +647,30 @@ module MD = struct
     let pci_address = Db.PCI.get_pci_id ~__context ~self:pci in
     Xenops_interface.Pci.address_of_string pci_address
 
+  let get_virtual_pci_address ~__context vgpu =
+   let device = vgpu.Db_actions.vGPU_device in
+   let virtual_pci_address  =  Xapi_globs.vgpu_pci_prefix ^ device ^".0" in
+   Xenops_interface.Pci.address_of_string virtual_pci_address
+
   let of_nvidia_vgpu ~__context vm vgpu =
     let open Vgpu in
     (* Get the PCI address. *)
     let physical_pci_address = get_target_pci_address ~__context vgpu in
-    (* Get the vGPU config. *)
+    let virtual_pci_address = get_virtual_pci_address ~__context vgpu in
+    (* Get the vGPU type_id, since psubdev_id is option, can't decide the string format *)
     let vgpu_type = vgpu.Db_actions.vGPU_type in
-    let internal_config =
-      Db.VGPU_type.get_internal_config ~__context ~self:vgpu_type in
-    let config_file =
-      try List.assoc Xapi_globs.vgpu_config_key internal_config
-      with Not_found -> failwith "NVIDIA vGPU config file not specified"
-    in
-    let config_file =
+    let identifier = Db.VGPU_type.get_identifier ~__context ~self:vgpu_type in
+    let type_id =
       try
-        let extra_args =
-          List.assoc Xapi_globs.vgpu_extra_args_key vm.API.vM_platform in
-        Printf.sprintf "%s,%s" config_file extra_args
-      with Not_found -> config_file
+        Scanf.sscanf identifier "%04d,nvidia,%04x,%04x,%04x,%04x,%s" (fun _ _ _ _ _ type_id -> type_id)
+      with _ ->
+        Scanf.sscanf identifier "%04d,nvidia,%04x,,%04x,%04x,%s" (fun _ _ _ _ type_id -> type_id)
     in
     let implementation =
       Nvidia {
         physical_pci_address = None; (* unused *)
-        config_file;
```

Comment:
Removing this field is likely to break the upgrade case.

Reply:
Sure, I will add this back here and IDL.
And I am a little concern about during upgrade/migrate case, as we keep the old field and adding new fields, can you point where to handle the new fields? I am wondering this is a common way for xenopsd handle new things.

Reply:
It may be possible to handle this in xenopsd, in the `import_metadata` function in xenops_server.ml. You need to do something special if `config_file = None`. The `virtual_pci_address` can just be set to `0000:00:0a.0` in that case, but I don't know where you'll get the `type_id` from....

Reply:
@robhoes , in upgrade scenario, can we keep the handling of type_id and config_file fields in XAPI in this way?
When the metadata is imported, we check "config_file", if it isn't none, in xapi we find the corresponding type_id, and allocate the virtual_pci_address (or write a hardcode "0000:00:0a.0"). By parsing the conf_file file name, we can find the matching type_id.
Then we can guarantee XAPI always send valid type_id and virtual_pci_address to Xenopsd.